### ISSUE TITLE
Implement fully satisfactory out-of-sync consensus prevention

### ISSUE THREAD

**Issue by lorenzo-dev1** [autonity/autonity#939]
We do not want to start the consensus engine before we are fully synced with the network. If we start consensus while out of sync it can be useless at best, or cause equivocation for the local node at worst.

A first temporary solution is implemented in https://github.com/autonity/autonity/pull/927, but it is not good enough.

In particular this temporary solution can still fail us if we cannot to a subset of peers that are out of sync with the global chain head.

We want to implement a solution that:
- gives us high guarantee that we are indeed synced with the chain head (possibly relying on the ACN network and the BFT honesty assumption).
- It is resilient to malicious nodes advertising heights >> actual global chain height

Solving this issue might require changing the selection of the peer with which to sync with (not the one advertising the highest high, but the one which advertise an height that makes sense with other 2/3 of the nodes).

We also need to **list out all the possible syncing scenarios**, and analyze the implemented solution against all of them.

**Comment by tbssajal** [comment]
> It is resilient to malicious nodes advertising heights >> actual global chain height

In case we have information from more than 2/3 of the committee members, taking the median of their heights instead of max height should be enough I guess.

**Comment by lorenzo-dev1** [comment]
>> It is resilient to malicious nodes advertising heights >> actual global chain height

> In case we have information from more than 2/3 of the committee members, taking the median of their heights instead of max height should be enough I guess.

Yeah that could work I believe, however in the end I implemented a simpler solution in https://github.com/autonity/autonity/pull/1029. I basically just suspend malicious syncers for 30 seconds from when we fail the sync. This way we will have a time-window where we can try to sync from other peers.

> gives us high guarantee that we are indeed synced with the chain head (possibly relying on the ACN network and the BFT honesty assumption).

This part of the issue still remains unsolved so I will keep this issue open. I think we probably need the WAL to solve this one.
